File: C++/DesignPattern/CreationalDesignPattern/ProxyPatternDemo_cache.cpp
#include <iostream>
#include <memory>
#include <string>
#include <unordered_map>
using namespace std;

struct IRateLimitinStrategy {
    virtual ~IRateLimitinStrategy() = default;
};

struct Notification {
    string priority;
    string message;
};

// Subject
struct Database {
    virtual string fetchData(const string& query) = 0;
    virtual ~Database() = default;
};

// Real Subject
struct RealDatabase : Database {
    string fetchData(const string& query) override {
        cout << "Fetching confidential employee salary data...\n";
        return "result for '" + query + "'";
    }
};

// Proxy
class DatabaseProxy : public Database {
    string role;
    unique_ptr<RealDatabase> database;
    unordered_map<string, string> cache;

public:
    explicit DatabaseProxy(const string& role) : role(role) {}

    string fetchData(const string& query) override {
        auto it = cache.find(query);
        if (it != cache.end()) {
            return it->second;
        }

        if (!database) {
            database = make_unique<RealDatabase>();
        }

        string value = database->fetchData(query);
        cache[query] = value;
        return value;
    }
};

int main() {
    unique_ptr<Database> user = make_unique<DatabaseProxy>("");
    cout << user->fetchData("dsadsadasd") << "\n";
    return 0;
}

File: Python/DesignPattern/CreationalDesignPattern/ProxyPatternDemo_cache.py
from abc import ABC, abstractmethod


class IRateLimitinStrategy(ABC):
    pass


class Notification:
    def __init__(self, priority: str = "", message: str = ""):
        self.priority = priority
        self.message = message


class Database(ABC):
    @abstractmethod
    def fetch_data(self, query: str) -> str: pass


class RealDatabase(Database):
    def fetch_data(self, query: str) -> str:
        print("Fetching confidential employee salary data...")
        return f"result for '{query}'"


class DatabaseProxy(Database):
    def __init__(self, role: str):
        self.role = role
        self.database = None
        self.cache = {}

    def fetch_data(self, query: str) -> str:
        if query in self.cache:
            return self.cache[query]

        if self.database is None:
            self.database = RealDatabase()

        value = self.database.fetch_data(query)
        self.cache[query] = value
        return value


if __name__ == "__main__":
    user = DatabaseProxy(None)
    print(user.fetch_data("dsadsadasd"))
